### PR TITLE
[Enhancement] Pass localNode info to all plugins on node start

### PR DESCRIPTION
**Pull request by DarshitChanpura**
Updates Node start to now pass localNode info so that it is available directly to a plugin instead of hack-y implementation

This deficit was encountered while trying to solve https://github.com/opensearch-project/security/issues/2724 which required localNode information in order to check whether the intercepted requests were for/from localNode.
The approaches tried:
1. Modifying GuiceHolder to holder localNode value fetched from TransportService. <- unreliable.
2. Using ClusterState's `getLocalNode()` <- fails with AssertionError stating that the calling class is not an instance of ClusterStateObserver.
3. Using ClusterService's localNode() method in OpenSearchSecurityPlugin class's `onNodeStarted()` implementation <- doesn't contain any value.

And so with recommendation from @reta , implemented this approach.

### Related Issues
https://github.com/opensearch-project/security/issues/2724

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/plugins/ClusterPlugin.java [hunk 1]
```diff
@@ -89,4 +91,15 @@ default Map<String, ExistingShardsAllocator> getExistingShardsAllocators() {
      */
     default void onNodeStarted() {}
 
+    /**
+     * Called when node is started. DiscoveryNode argument is passed to allow referring localNode value inside plugin
+     * NOTE: Currently this function implementation is limited to OpenSearchSecurityPlugin inside {@link Node#start()}
+     *
+     * TODO: modify javadoc when the restriction to SecurityPlugin is lifted
+     * @param localNode Current Node info
```
Comment (reta) [inline server/src/main/java/org/opensearch/plugins/ClusterPlugin.java:99]:
```suggestion
     * @param localNode local Node info
```

**Comment by reta** [comment]
@dblock @andrross curious to hear you opinion, it seems to be completing the callback with missing contextual information (backward compatible change).

File: server/src/main/java/org/opensearch/plugins/ClusterPlugin.java [hunk 2]
```diff
@@ -89,4 +90,13 @@ default Map<String, ExistingShardsAllocator> getExistingShardsAllocators() {
      */
     default void onNodeStarted() {}
```
Comment (peternied) [inline server/src/main/java/org/opensearch/plugins/ClusterPlugin.java:91]: What do you think of marking the `onNodeStarted()` as deprecated so developers will know which one to use?

**Review by peternied — APPROVED** [review]

File: server/src/main/java/org/opensearch/plugins/ClusterPlugin.java (hunk 2 above)
Comment (DarshitChanpura) [reply server/src/main/java/org/opensearch/plugins/ClusterPlugin.java:91]: Sounds like a great idea.

**Comment by reta** [comment]
@andrross @dblock any risks you see with this change?

**Comment by DarshitChanpura** [comment]
@reta @dblock @peternied Could you help get this merged? It will unblock https://github.com/opensearch-project/security/pull/2765

**Review by dblock — DISMISSED** [review]
Add a test that exercises this code to avoid regressions. When a node start, I expect `onNodeStarted` to be called with the correct node.

**Comment by DarshitChanpura** [comment]
> Add a test that exercises this code to avoid regressions. When a node start, I expect `onNodeStarted` to be called with the correct node.

Done.

File: server/src/test/java/org/opensearch/plugins/ClusterPluginTests.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+import java.util.Collection;
+
+public class ClusterPluginTests extends OpenSearchSingleNodeTestCase {
+
+    @Override
+    protected Collection<Class<? extends Plugin>> getPlugins() {
+        return pluginList(DummyClusterPlugin.class);
+    }
+
+    public void testOnNodeStarted_shouldContainLocalNodeInfo() {
```
Comment (peternied) [inline server/src/test/java/org/opensearch/plugins/ClusterPluginTests.java:23]: Thanks for the tests, could you what about adding a test that confirms `onNodeStarted()` is called on plugins that aren't using the newer interface?

File: server/src/main/java/org/opensearch/plugins/ClusterPlugin.java [hunk 4]
```diff
@@ -87,6 +88,16 @@ default Map<String, ExistingShardsAllocator> getExistingShardsAllocators() {
     /**
      * Called when the node is started
      */
+    @Deprecated
```
Comment (peternied) [inline server/src/main/java/org/opensearch/plugins/ClusterPlugin.java:91]: Nit: could you include instructions to use the other `onNodeStarted(DiscoveryNode)` instead?

**Review by peternied — APPROVED** [review]
Thanks for adding these tests, looks good to me

**Review by cwperks — APPROVED** [review]

**Review by saratvemulapalli — APPROVED** [review]

**Comment by reta** [comment]
@DarshitChanpura sadly needs manual backport to `2.x`

**Pull request merged**